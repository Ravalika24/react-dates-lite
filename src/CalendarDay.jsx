/* @flow */
import * as React from 'react';
import styled from 'styled-components';

type ButtonProps = {
  isPast: boolean,
  isFuture: boolean,
  isHidden: boolean,
  isHovered: boolean,
  isSelected: boolean,
  isDisabled: boolean,
  colors: {| [string]: string |}
};

type Props = ButtonProps & {
  value: Date,
  number: number,
  selectDate: Date => void,
  onHover: Date => void,
  // eslint-disable-next-line react/no-unused-prop-types
  classes: {| [string]: string |}
};

export const getClasses = (props: Props) => {
  if (props.isPast || props.isFuture || props.isDisabled) {
    return props.classes.disabled;
  }
  if (props.isHovered) {
    return props.classes.hovered;
  }
  if (props.isSelected) {
    return props.classes.selected;
  }

  return props.classes.date;
};

export const buttonColor = (props: ButtonProps) => {
  if (props.className) {
    return 'inherit';
  }
  if (props.isPast || props.isFuture) {
    return 'inherit';
  }
  if (props.isHovered) {
    return 'white';
  }
  if (props.isSelected) {
    return 'white';
  }

  return 'inherit';
};

const buttonBg = (props: ButtonProps) => {
  if (props.className) {
    return 'inherit';
  }
  if (props.isPast || props.isFuture) {
    return 'white';
  }

  if (props.isHovered) {
    return props.colors.hovered;
  }
  if (props.isSelected) {
    return props.colors.selected;
  }

  return 'white';
};

const Button = styled.button`
  width: 100%;
  height: 40px;
  display: flex;
  align-items: center;
  justify-content: center;
  opacity: ${props => (props.isHidden ? 0 : 1)};
  color: ${props => buttonColor(props)};
  background: ${props => buttonBg(props)};
  border: none;
  cursor: pointer;
  border-radius: 0;
  :hover {
    background: ${props =>
      props.isHovered ? props.colors.hovered : props.colors.hover};
  }
  :disabled {
    cursor: default;
    :hover {
      background: inherit;
    }
  }
  :focus {
    outline: none;
  }
`;

const Td = styled.div`
  display: table-cell;
  border: ${props => (props.isHidden ? 'none' : '1px solid #e4e7e7')};
  box-sizing: border-box;
`;

export default class CalendarDay extends React.PureComponent<Props> {
  props: Props;

  handleClick = () => {
    this.props.selectDate(this.props.value);
  };

  handleHover = () => {
    this.props.onHover(this.props.value);
  };

  render() {
    const {
      isHidden,
      number,
      isPast,
      isFuture,
      isSelected,
      isDisabled,
      isHovered,
      colors
    } = this.props;

    return (
      <Td isHidden={isHidden}>
        <Button
          className={getClasses(this.props)}
          data-test="CalendarDay"
          onClick={this.handleClick}
          onMouseOver={this.handleHover}
          onFocus={this.handleHover}
          disabled={isPast || isFuture || isDisabled}
          isSelected={isSelected}
          isHidden={isHidden}
          isPast={isPast}
          isFuture={isFuture}
          isHovered={isHovered}
          colors={colors}>
          {number}
        </Button>
      </Td>
    );
  }
}
