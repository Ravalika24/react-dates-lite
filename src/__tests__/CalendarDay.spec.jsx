import React from 'react';
import { shallow } from 'enzyme';

import getDate from 'date-fns/get_date';

import CalendarDay from '../CalendarDay';

describe('#CalendarDay', () => {
  const date = new Date(2018, 1, 1);

  it('should render correctly', () => {
    const wrapper = shallow(
      <CalendarDay
        isHidden={false}
        number={getDate(date)}
        value={date}
        selectDate={jest.fn()}
        onHover={jest.fn()}
        isHovered={false}
        isSelected={false}
        isPast={false}
        isFuture={false}
        colors={{}}
        classes={{}}
      />
    );
    expect(wrapper.getElement()).toMatchSnapshot();
  });
  it('should render hidden', () => {
    const wrapper = shallow(
      <CalendarDay
        isHidden
        number={getDate(date)}
        value={date}
        selectDate={jest.fn()}
        onHover={jest.fn()}
        isHovered={false}
        isSelected={false}
        isPast={false}
        isFuture={false}
        colors={{}}
        classes={{}}
      />
    );
    expect(wrapper.getElement()).toMatchSnapshot();
  });

  it('should handleHover', () => {
    const onHover = jest.fn();
    const wrapper = shallow(
      <CalendarDay
        isHidden={false}
        number={getDate(date)}
        value={date}
        selectDate={jest.fn()}
        onHover={onHover}
        isHovered={false}
        isSelected={false}
        isPast={false}
        isFuture={false}
        colors={{}}
        classes={{}}
      />
    );

    const node = wrapper.instance();
    node.handleHover();

    expect(onHover).toBeCalledWith(date);
  });

  it('should handleClick', () => {
    const onClick = jest.fn();
    const wrapper = shallow(
      <CalendarDay
        isHidden={false}
        number={getDate(date)}
        value={date}
        selectDate={onClick}
        onHover={jest.fn()}
        isHovered={false}
        isSelected={false}
        isPast={false}
        isFuture={false}
        colors={{}}
        classes={{}}
      />
    );

    const node = wrapper.instance();
    node.handleClick();

    expect(onClick).toBeCalledWith(date);
  });
});
