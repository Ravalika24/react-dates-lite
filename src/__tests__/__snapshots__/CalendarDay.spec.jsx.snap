// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#CalendarDay should render correctly 1`] = `
<styled.div
  isHidden={false}
>
  <styled.button
    className={undefined}
    colors={Object {}}
    data-test="CalendarDay"
    disabled={undefined}
    isFuture={false}
    isHidden={false}
    isHovered={false}
    isPast={false}
    isSelected={false}
    onClick={[Function]}
    onFocus={[Function]}
    onMouseOver={[Function]}
  >
    1
  </styled.button>
</styled.div>
`;

exports[`#CalendarDay should render hidden 1`] = `
<styled.div
  isHidden={true}
>
  <styled.button
    className={undefined}
    colors={Object {}}
    data-test="CalendarDay"
    disabled={undefined}
    isFuture={false}
    isHidden={true}
    isHovered={false}
    isPast={false}
    isSelected={false}
    onClick={[Function]}
    onFocus={[Function]}
    onMouseOver={[Function]}
  >
    1
  </styled.button>
</styled.div>
`;
